pkg/clusters/addons/kong/kong.go

```diff
@@ -0,0 +1,266 @@
+package kong
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os/exec"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters"
+)
+
+// -----------------------------------------------------------------------------
+// Kong Addon
+// -----------------------------------------------------------------------------
+
+const (
+	// AddonName is the unique name of the Kong cluster.Addon
+	AddonName clusters.AddonName = "kong"
+)
+
+// Addon is a Kong Proxy addon which can be deployed on a clusters.Cluster.
+type Addon struct {
+	namespace  string
+	name       string
+	deployArgs []string
+	dbmode     DBMode
+}
+
+// New produces a new clusters.Addon for Kong but uses a very opionated set of
+// default configurations (see the defaults() function for more details).
+// If you need to customize your Kong deployment, use the kong.Builder instead.
+func New() *Addon {
+	return NewBuilder().Build()
+}
+
+// ProxyURL provides a routable *url.URL for accessing the Kong proxy.
+func (a *Addon) ProxyURL(ctx context.Context, cluster clusters.Cluster) (*url.URL, error) {
+	waitForObjects, ready, err := a.Ready(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ready {
+		return nil, fmt.Errorf("the addon is not ready on cluster %s, see: %+v", cluster.Name(), waitForObjects)
+	}
+
+	service, err := cluster.Client().CoreV1().Services(a.namespace).Get(ctx, DefaultProxyServiceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return url.Parse(fmt.Sprintf("http://%s:%d", service.Status.LoadBalancer.Ingress[0].IP, DefaultProxyTCPServicePort))
+	}
+
+	return nil, fmt.Errorf("service %s has not yet been provisoned", service.Name)
+}
+
+// ProxyAdminURL provides a routable *url.URL for accessing the Kong Admin API.
+func (a *Addon) ProxyAdminURL(ctx context.Context, cluster clusters.Cluster) (*url.URL, error) {
```

Comment:
These ProxyXXXURL functions appear very similar, looks like they do the same thing and only diverge in details like the name of the service.

Can they be reduced into one reusable implementation?

Reply:
Sure, refactored in 95aa855cf7ea58610a5f6a46de70cb834813135d